src/visualization/interaction/Highlighter.js

```diff
@@ -35,62 +35,42 @@ ROS3D.Highlighter.prototype.onMouseOver = function(event) {
  * @param event - the event that contains the target of the mouseout
  */
 ROS3D.Highlighter.prototype.onMouseOut = function(event) {
-  this.hoverObjs.splice(this.hoverObjs.indexOf(event.currentTarget), 1);
+  this.highlightObject(event.currentTarget, false);
 };
 
+
 /**
- * Add all corresponding webgl objects in the given scene and add them to the given render list.
+ * Highlight and unhighlight the given object
  *
- * @param scene - the scene to check for webgl objects
- * @param objects - the objects list to check
- * @param renderList - the list to add to
+ * @param target_object - the target object to (un)highlight
+ * @param highlight - whether to highlight or unhighlight
  */
-ROS3D.Highlighter.prototype.getWebglObjects = function(scene, objects, renderList) {
-  var objlist = scene.__webglObjects;
-  // get corresponding webgl objects
-  for ( var c = 0; c < objects.length; c++) {
-    if (objects[c]) {
-      for ( var o = objlist.length - 1; o >= 0; o--) {
-        if (objlist[o].object === objects[c]) {
-          renderList.push(objlist[o]);
-          break;
-        }
+ROS3D.Highlighter.prototype.highlightObject = function (target_object, highlight) {
+  if(target_object.material === undefined) {
+    if(target_object.children === undefined) {
+      return;
+    }
+    else {
+      for(var c = 0 ; c < target_object.children.length; c++) {
+        this.highlightObject(target_object.children[c], highlight);
       }
-      // recurse into children
-      this.getWebglObjects(scene, objects[c].children, renderList);
     }
   }
-};
-
-/**
- * Render highlighted objects in the scene.
- *
- * @param renderer - the renderer to use
- * @param scene - the scene to use
- * @param camera - the camera to use
- */
-ROS3D.Highlighter.prototype.renderHighlight = function(renderer, scene, camera) {
-  // get webgl objects
-  var renderList = [];
-  this.getWebglObjects(scene, this.hoverObjs, renderList);
-
-  // define highlight material
-  scene.overrideMaterial = new THREE.MeshBasicMaterial({
-    fog : false,
-    opacity : 0.5,
-    depthTest : true,
-    depthWrite : false,
-    polygonOffset : true,
-    polygonOffsetUnits : -1,
-    side : THREE.DoubleSide
-  });
-
-  // swap render lists, render, undo
-  var oldWebglObjects = scene.__webglObjects;
-  scene.__webglObjects = renderList;
-
-  renderer.render(scene, camera);
-
-  scene.__webglObjects = oldWebglObjects;
-  scene.overrideMaterial = null;
+  else {
+    if(highlight) {
+      target_object.originalMaterial = target_object.material;
+      target_object.material = new THREE.MeshBasicMaterial({
+        fog : false,
+        opacity : 0.5,
+        depthTest : true,
+        depthWrite : false,
+        polygonOffset : true,
+        polygonOffsetUnits : -1,
+        side : THREE.DoubleSide
+      });
+    }
+    else {
+      target_object.material = target_object.originalMaterial;
```

Comment:
One more thing: We should probably be defensive here and check for the existence of ```originalMaterial```.

If ```highlightObject(false)``` is ever called on an object where ```highlight(true)``` has not been called, we'd be in trouble!

Reply:
Yeah, I agree. I would improve this logic with another PR.

Reply:
Please don't. Merging this PR as-is would introduce a serious bug (see the screen shot I attached to the issue), and fixing it would be trivial (just give the existing material an emissive component instead of switching the material completely).

Reply:
It is not as easy as you described and it doesn't introduce serious bug while I was testing. This PR is already pretty large and getting longer and I am running out of time. Since it is already a big change which covers most of use cases. I will move on and start to patch small stuffs. I don't see any serious reason that this should not go into the develop branch.